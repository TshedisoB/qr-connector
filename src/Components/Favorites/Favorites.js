import React, { useState, useEffect } from "react";
import { View, Text, TouchableOpacity } from "react-native";
import { AntDesign } from "@expo/vector-icons";

import { updateFavorites, getAllContacts } from "../../sqlQueries/sqlQueries";
import { checkBoolValue } from "../../utils/helper";
import { styles } from "./Favorites.style";
import { db } from "../ScanQr/ScanQr";

const Favorites = ({ contact }) => {
  const [data, setData] = useState([]);
  const [isFavorite, setIsFavorite] = useState();
  const [favoriteBoolStatus, setFavoriteBoolStatus] = useState();

  useEffect(() => {
    getAllContacts(db, (data) => {
      setData(data);
    });

    const scannedEmail = contact[0].email;
    const matchingUser = data.find((user) => user.email === scannedEmail);

    if (matchingUser) {
      setIsFavorite(matchingUser.isFavorite);
    }

    setFavoriteBoolStatus(checkBoolValue(isFavorite));
  }, [data, isFavorite, favoriteBoolStatus]);

  const handleAddToFavorites = () => {
    const dbEmails = [];
    let email = contact[0].email;

    data.forEach((data) => {
      dbEmails.push(data.email);
    });

    if (dbEmails.includes(email)) {
      updateFavorites(db, email, favoriteBoolStatus);
    }
  };

  return (
    <View style={styles.container}>
      <TouchableOpacity
        onPress={handleAddToFavorites}
        style={styles.button}
        testID="favorite-status">
        {favoriteBoolStatus ? (
          <>
            <Text style={styles.buttonText}>Remove from favorites</Text>
            <AntDesign name="star" style={styles.antIcon} />
          </>
        ) : (
          <>
            <Text style={styles.buttonText}>Add to favorites</Text>
            <AntDesign name="staro" style={styles.antIconInactive} />
          </>
        )}
      </TouchableOpacity>
    </View>
  );
};

export default Favorites;
