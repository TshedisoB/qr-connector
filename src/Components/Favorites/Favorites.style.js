import { StyleSheet } from "react-native";

export const styles = StyleSheet.create({
  container: {
    alignItems: "center",
    margin: 4,
  },
  button: {
    alignItems: "center",
    backgroundColor: "#eee",
    padding: 10,
    borderRadius: 5,
  },
  buttonText: {
    marginRight: 10,
  },
  antIcon: {
    color: "yellow",
    fontSize: 25,
  },
  antIconInactive: {
    fontSize: 20,
  },
});
