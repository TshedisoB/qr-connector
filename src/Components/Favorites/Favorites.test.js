import React from "react";
import { render } from "@testing-library/react-native";

import Favorites from "./Favorites";

jest.mock("../../sqlQueries/sqlQueries", () => ({
  __esModule: true,
  updateFavorites: jest.fn(),
  getAllContacts: jest.fn(),
}));

describe("Favorites component", () => {
  const contact = [
    {
      name: "Tshediso Boshiana",
      email: "tshediso.boshiana@umuzi.org",
    },
  ];

  it("should render correctly", () => {
    const tree = render(<Favorites contact={contact} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should display text 'Add to favorites' after scanning a QR code", () => {
    const { getByText } = render(<Favorites contact={contact} />);

    expect(getByText("Add to favorites")).toBeTruthy();
  });
});
