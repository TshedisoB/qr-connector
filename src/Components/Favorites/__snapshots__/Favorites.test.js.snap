// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites component should render correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "margin": 4,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#eee",
        "borderRadius": 5,
        "opacity": 1,
        "padding": 10,
      }
    }
    testID="favorite-status"
  >
    <Text
      style={
        {
          "marginRight": 10,
        }
      }
    >
      Add to favorites
    </Text>
    <Text />
  </View>
</View>
`;
